pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go

```diff
@@ -237,3 +243,238 @@ func TestDeleteAddressWithWrongTier(t *testing.T) {
+		return nil, err
+	}
+
+	nodes := []*v1.Node{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nodeName,
+				Labels: map[string]string{
+					kubeletapis.LabelHostname: nodeName,
+				},
+			},
+			Status: v1.NodeStatus{
+				NodeInfo: v1.NodeSystemInfo{
+					KubeProxyVersion: "v1.7.2",
+				},
+			},
+		},
+	}
+
+	status, err := gce.ensureExternalLoadBalancer(
+		clusterName,
+		clusterID,
+		apiService,
+		nil,
+		nodes,
+	)
+
+	return status, err
+}
+
+func TestEnsureExternalLoadBalancer(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+
+	status, err := createExternalLoadBalancer(gce)
+	assert.NoError(t, err)
+	assert.NotEmpty(t, status.Ingress)
+
+	lbName := cloudprovider.GetLoadBalancerName(apiService)
+
+	// Check that Firewall is created
+	firewall, err := gce.GetFirewall(MakeFirewallName(lbName))
+	require.NoError(t, err)
+	assert.Equal(t, []string{nodeName}, firewall.TargetTags)
+	assert.Equal(t, []string{"0.0.0.0/0"}, firewall.SourceRanges)
+
+	// Check that TargetPool is Created
+	pool, err := gce.GetTargetPool(lbName, gceRegion)
+	require.NoError(t, err)
+	assert.Equal(t, lbName, pool.Name)
+	assert.NotEmpty(t, pool.HealthChecks)
+	assert.Equal(t, 1, len(pool.Instances))
+
+	// Check that HealthCheck is created
+	hcName := MakeNodesHealthCheckName(clusterID)
+	healthcheck, err := gce.GetHttpHealthCheck(hcName)
+	require.NoError(t, err)
+	assert.Equal(t, hcName, healthcheck.Name)
+
+	// Check that ForwardingRule is created
+	fwdRule, err := gce.GetRegionForwardingRule(lbName, gceRegion)
+	require.NoError(t, err)
+	assert.Equal(t, lbName, fwdRule.Name)
+	assert.Equal(t, "TCP", fwdRule.IPProtocol)
+	assert.Equal(t, "123-123", fwdRule.PortRange)
+}
+
+func TestUpdateExternalLoadBalancer(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+
+	_, err = createExternalLoadBalancer(gce)
+	assert.NoError(t, err)
+
+	newNodeName := "test-node-2"
+	gce.InsertInstance(
+		gceProjectId,
+		zoneName,
+		&compute.Instance{
+			Name: newNodeName,
+			Tags: &compute.Tags{
+				Items: []string{newNodeName},
+			},
+		},
+	)
+
+	newNodes := []*v1.Node{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nodeName,
+				Labels: map[string]string{
+					kubeletapis.LabelHostname: nodeName,
+				},
+			},
+			Status: v1.NodeStatus{
+				NodeInfo: v1.NodeSystemInfo{
+					KubeProxyVersion: "v1.7.2",
+				},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: newNodeName,
+				Labels: map[string]string{
+					kubeletapis.LabelHostname: newNodeName,
+				},
+			},
+			Status: v1.NodeStatus{
+				NodeInfo: v1.NodeSystemInfo{
+					KubeProxyVersion: "v1.7.2",
+				},
+			},
+		},
+	}
+
+	err = gce.updateExternalLoadBalancer(clusterName, apiService, newNodes)
+	assert.NoError(t, err)
+
+	lbName := cloudprovider.GetLoadBalancerName(apiService)
+
+	// Check that TargetPool is updated with the new node
+	pool, err := gce.GetTargetPool(lbName, gceRegion)
+	require.NoError(t, err)
+
+	assert.Equal(
```

Comment:
We probably shouldn't assert order. Unfortunately, looks like we're using an old version of testify which doesn't have `ElementsMatch`.  There may be a util func somewhere that checks string set equality.

Reply:
Added a TODO to use ElementsMatch when we upgrade, but for now use `assert.Contains`.